src/polly/src/epoll.rs

```diff
@@ -0,0 +1,201 @@
+    pub fn new(events: EventType, data: u64) -> Self {
+        Event(epoll_event {
+            events: events.bits(),
+            u64: data,
+        })
+    }
+
+    pub fn events(&self) -> u32 {
+        self.events
+    }
+
+    pub fn data(&self) -> u64 {
+        self.u64
+    }
+}
+
+/// Stores a file descriptor referring to an epoll instance.
+#[derive(Clone, Copy, Debug)]
+pub struct Epoll {
+    epoll_fd: RawFd,
+}
+
+impl Epoll {
+    /// Create a new epoll file descriptor.
+    pub fn new() -> io::Result<Self> {
+        let epoll_fd = SyscallReturnCode(unsafe { epoll_create1(EPOLL_CLOEXEC) }).into_result()?;
+        Ok(Epoll { epoll_fd })
+    }
+
+    /// Wrapper for `libc::epoll_ctl`.
+    ///
+    /// This can be used for adding, modifying or removing a file descriptor in the
+    /// interest list of the epoll instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `operation` refers to the action to be performed on the file descriptor.
+    /// * `fd` is the file descriptor on which we want to perform `operation`.
+    /// * `event` refers to the `epoll_event` instance that is linked to `fd`.
+    pub fn ctl(self, operation: ControlOperation, fd: RawFd, mut event: Event) -> io::Result<()> {
+        // Safe because we give a valid epoll file descriptor, a valid file descriptor to watch,
+        // as well as a valid epoll_event structure. We also check the return value.
+        SyscallReturnCode(unsafe {
+            epoll_ctl(
+                self.epoll_fd,
+                operation as i32,
+                fd,
+                &mut event as *mut _ as *mut epoll_event,
+            )
+        })
+        .into_empty_result()
+    }
+
+    /// Wrapper for `libc::epoll_wait`.
+    ///
+    /// # Arguments
+    ///
+    /// * `events` points to a memory area that will be used for storing the events
+    /// returned by `epoll_wait()` call.
+    /// * `max_events` is the maximum number of events that we want to be stored in
+    /// `events` buffer.
+    /// * `timeout` specifies for how long the `epoll_wait` system call will block
+    /// (measured in milliseconds).
+    pub fn wait(self, events: &mut [Event], max_events: usize, timeout: i32) -> io::Result<usize> {
+        // Safe because we give a valid epoll file descriptor and an array of epoll_event structures
+        // that will be modified by the kernel to indicate information about the subset of file
+        // descriptors in the interest list. We also check the return value.
+        let events_count = SyscallReturnCode(unsafe {
+            epoll_wait(
+                self.epoll_fd,
+                events.as_mut_ptr() as *mut epoll_event,
+                max_events as i32,
+                timeout,
+            )
+        })
+        .into_result()? as usize;
+
+        Ok(events_count)
+    }
+}
+
+impl AsRawFd for Epoll {
+    fn as_raw_fd(&self) -> RawFd {
+        self.epoll_fd
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
We need more tests here. While most of the functionality is tested in the event_manager, we should still have Epoll tests here because when we are going to move this to rust-vmm those code paths will not be exercised any more.